{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import time\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def load_fasta(filename, verbose=0):\n",
    "    \"\"\"\n",
    "    Parses a classically formatted and possibly\n",
    "    compressed FASTA file into a dictionary where the key\n",
    "    for a sequence is the first part of its header without\n",
    "    any white space; if verbose is nonzero then the identifiers\n",
    "    together with lengths of the read sequences are printed\n",
    "    \"\"\"\n",
    "    if filename.endswith(\".gz\"):\n",
    "        fp = gzip.open(filename, 'rt')\n",
    "    else:\n",
    "        fp = open(filename, 'r')\n",
    "    # split at headers\n",
    "    # data = fp.read().split('>')\n",
    "    data = fp.read()\n",
    "    data = data.split('>')\n",
    "    fp.close()\n",
    "    # ignore whatever appears before the 1st header\n",
    "    data.pop(0)\n",
    "    # prepare the dictionary\n",
    "    D = {}\n",
    "    for sequence in data:\n",
    "        lines = sequence.split('\\n')\n",
    "        header = lines.pop(0).split()\n",
    "        key = header[0]\n",
    "        D[key] = ''.join(lines)\n",
    "        if verbose:\n",
    "            print(\"Sequence %s of length %d read\" % (key, len(D[key])))\n",
    "    return D\n",
    "\n",
    "\n",
    "nucleotides = {'A': 0, 'C': 1, 'G': 2, 'T': 3}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def char_diff(s1, s2):\n",
    "    \"\"\"\n",
    "\n",
    "    :param s1: first string sequence\n",
    "    :param s2: second string sequence\n",
    "    :return: number of pairwise different characters between string s1 and string s2\n",
    "    \"\"\"\n",
    "    err = 0\n",
    "    for c1, c2 in zip(s1, s2):\n",
    "        if c1 != c2:\n",
    "            err += 1\n",
    "    return err\n",
    "\n",
    "\n",
    "def score(motifs):\n",
    "    \"\"\"\n",
    "    :param motifs: a list of motifs, where each motif is represented as a string\n",
    "    :return: sum of number of different characters between all possible pairs of motifs\n",
    "    \"\"\"\n",
    "    err = 0\n",
    "    d = len(motifs)\n",
    "\n",
    "    for i in range(d):\n",
    "        for j in range(i + 1, d):\n",
    "            err += char_diff(motifs[i], motifs[j])\n",
    "\n",
    "    return err\n",
    "\n",
    "\n",
    "def connected_components(motifs, mutations):\n",
    "    \"\"\"\n",
    "\n",
    "    :param motifs: a list of motifs, where each motif is represented as a string\n",
    "    :param mutations: integer of allowed mutations in each motif\n",
    "    :return: boolean if motifs can form a connected graph\n",
    "    \"\"\"\n",
    "\n",
    "    d = len(motifs)\n",
    "    adjacency_matrix = np.zeros((d, d))\n",
    "    for i in range(d):\n",
    "        for j in range(d):\n",
    "            adjacency_matrix[i, j] = char_diff(motifs[i], motifs[j]) <= mutations\n",
    "\n",
    "    G = nx.from_numpy_matrix(adjacency_matrix)\n",
    "    return nx.number_connected_components(G)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_profile(motifs, beta=1):\n",
    "    \"\"\"\n",
    "\n",
    "    :param motifs: a list of motifs, where each motif is represented as a string\n",
    "    :param beta: laplace correction value\n",
    "    :return: a profile of size lx4 where l is the length of a motif and 4 represents the number of amino acids\n",
    "    \"\"\"\n",
    "    l = len(motifs[0])\n",
    "    profile = np.zeros((l, 4))\n",
    "    profile += beta\n",
    "    for count in range(l):\n",
    "        for motif in motifs:\n",
    "            x = nucleotides[motif[count]]\n",
    "            profile[count][x] += 1\n",
    "\n",
    "    return profile / np.linalg.norm(profile, axis=1, ord=1).reshape((l, 1))\n",
    "\n",
    "\n",
    "def get_background(DNAs, l):\n",
    "    \"\"\"\n",
    "    \n",
    "    :param DNAs:  list of strings where each string represents a DNA\n",
    "    :param l: length of motif\n",
    "    :return: background probabilities profile Q \n",
    "    \"\"\"\n",
    "    t = len(DNAs)\n",
    "    Q = [np.zeros((l, 4)) for _ in range(t)]\n",
    "\n",
    "    for i in range(t):\n",
    "        for j in range(t):\n",
    "            if i != j:\n",
    "                for n in range(len(DNAs[j]) - l + 1):\n",
    "                    for m in range(l):\n",
    "                        ch = DNAs[j][n + m]\n",
    "                        Q[i][m][nucleotides[ch]] += 1\n",
    "        Q[i] /= np.linalg.norm(Q[i], axis=1, ord=1).reshape((l, 1))\n",
    "    return Q\n",
    "\n",
    "\n",
    "def new_motif(motifs, DNA, Q):\n",
    "    \"\"\"\n",
    "\n",
    "    :param motifs: a list of motifs, where each motif is represented as a string\n",
    "    :param DNA:  string\n",
    "    :return: newly randomly chosen motif, probability of motif being chose depends on  calculated weights\n",
    "    \"\"\"\n",
    "\n",
    "    l = len(motifs[0])\n",
    "    profile = get_profile(motifs)\n",
    "\n",
    "    weights = np.zeros(len(DNA) - l + 1)\n",
    "    for i in range(len(DNA) - l + 1):\n",
    "        pr = 1\n",
    "        pr_q = 1\n",
    "        for j in range(l):\n",
    "            x = nucleotides[DNA[i + j]]\n",
    "            pr *= profile[j][x]\n",
    "            pr_q *= Q[j][x]\n",
    "        weights[i] = pr/pr_q\n",
    "\n",
    "    weights = weights / np.sum(weights)\n",
    "    i = np.random.choice(list(range(len(weights))), p=weights)\n",
    "\n",
    "    return DNA[i:i + l]\n",
    "\n",
    "def gibbs_sampling(DNAs, l, n, mutations):\n",
    "    \"\"\"\n",
    "\n",
    "    :param DNAs: a list of strings where each string represents a DNA\n",
    "    :param l: length of motif\n",
    "    :param n: number of iterations of gibbs sampling\n",
    "    :param mutations:\n",
    "    :return: best found motifs\n",
    "    \"\"\"\n",
    "    t = len(DNAs)\n",
    "    optimal = False\n",
    "    Q = get_background(DNAs, l)\n",
    "\n",
    "    motifs = []\n",
    "    for DNA in DNAs:\n",
    "        i = np.random.randint(len(DNA) - l + 1)\n",
    "        motifs.append(DNA[i:i + l])\n",
    "\n",
    "    motifs = list(motifs)\n",
    "    best_score = score(motifs)\n",
    "    for _ in range(n):\n",
    "        h = np.random.randint(t)\n",
    "        other_motifs = motifs[0:h] + motifs[h + 1:t]\n",
    "        motifs[h] = new_motif(other_motifs, DNAs[h], Q[h])\n",
    "\n",
    "        if score(motifs) < best_score:\n",
    "            motifs = list(motifs)\n",
    "            best_score = score(motifs)\n",
    "            if connected_components(motifs, mutations) == 1:\n",
    "                optimal = True\n",
    "                break\n",
    "    return motifs, optimal\n",
    "\n",
    "\n",
    "def repeated_gibbs_sampling(DNAs, l, n=1000, repeats=20, mutations=4):\n",
    "    \"\"\"\n",
    "\n",
    "    :param DNAs: a list of strings where each string represents a DNA\n",
    "    :param l: length of motif\n",
    "    :param n: number of iterations of gibbs sampling\n",
    "    :param repeats: number of repeated gibbs samplings\n",
    "    :param mutations:\n",
    "    :return: best found motifs\n",
    "    \"\"\"\n",
    "\n",
    "    results, optimal = gibbs_sampling(DNAs, l, n, mutations)\n",
    "    best_score = score(results)\n",
    "    motifs = results[:]\n",
    "    for i in range(repeats - 1):\n",
    "        print(best_score)\n",
    "        if optimal:\n",
    "            break\n",
    "        results, optimal = gibbs_sampling(DNAs, l, n, mutations)\n",
    "\n",
    "        if score(results) < best_score:\n",
    "            best_score = score(results)\n",
    "            motifs = results[:]\n",
    "\n",
    "        \n",
    "    return motifs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "391\n",
      "200\n",
      "\n",
      "Obtained motifs:\n",
      "GTGACACATTATGTC\n",
      "GTTAGACAATATGTT\n",
      "GTGATATGTTATGAC\n",
      "CTGAAACACTATGTC\n",
      "CTGACACGTTATGTC\n",
      "GTGACACATTATGTC\n",
      "GAGACTCATAATGTC\n",
      "GGGACACATAATGTC\n",
      "GTGAGAGACGATGTC\n",
      "GTGACACATGGGGTC\n",
      "\n",
      "Time:13.050220489501953\n"
     ]
    }
   ],
   "source": [
    "\n",
    "l = 15\n",
    "D = load_fasta('MotSam.fasta')\n",
    "DNAs = []\n",
    "for d, DNA in D.items():\n",
    "    DNAs.append(DNA)\n",
    "start_time=time.time()\n",
    "results = repeated_gibbs_sampling(DNAs, l, mutations=4)\n",
    "end_time=time.time()-start_time\n",
    "\n",
    "print()\n",
    "print(\"Obtained motifs:\")\n",
    "for result in results:\n",
    "    print(result)\n",
    "print()\n",
    "print(\"Time:\"+str(end_time))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1. What motif of length 15 have you found and using which parameters of your program in the sequences from the file MotSam.fasta?\n",
    "\n",
    "I obtained the following motifs:  \n",
    "GTGATATGTTATGAC  \n",
    "GTGACACATTATGTC  \n",
    "GTTAGACAATATGTT  \n",
    "CTGAAACACTATGTC  \n",
    "CTGACACGTTATGTC  \n",
    "GTGACACATTATGTC  \n",
    "GTGAGAGACGATGTC  \n",
    "GAGACTCATAATGTC  \n",
    "GTGACACATGGGGTC  \n",
    "GGGACACATAATGTC  \n",
    "  \n",
    "Parameters:  \n",
    "- repeats of Gibbs sampling : 20  \n",
    "- number of iterations in Gibbs sampling: 1000  \n",
    "- Laplace correction beta: 1\n",
    "\n",
    "Laplace formula:  \n",
    "\n",
    "\\begin{equation}\n",
    "\\theta = \\cfrac{f_i+\\beta}{l+4\\beta},~~ i=\\{A,C,G,T\\}\n",
    "\\end{equation}  \n",
    "\n",
    "#### 2. What was the running time of your program and what can be seen during the execution of your program?\n",
    "\n",
    "Running time is usually under 2 minutes. We can see the number of diffrent characters between \"best\" motif of each sequence.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTGACACATTATGTC\n",
      "25\n"
     ]
    }
   ],
   "source": [
    "def motif_score(motifs, mutation):\n",
    "    scores = [0]*len(motifs)\n",
    "    for i, motif1 in enumerate(motifs):\n",
    "        for j, motif2 in enumerate(motifs):\n",
    "            err = char_diff(motif1, motif2)\n",
    "            if err>mutation:\n",
    "                scores[i] = float('Inf')\n",
    "                break\n",
    "            scores[i] += err\n",
    "    best_score_motif = min(scores)\n",
    "    return motifs[scores.index(best_score_motif)],best_score_motif\n",
    "\n",
    "motif, m_score = motif_score(results, 4)\n",
    "print(motif)\n",
    "print(m_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####  3. What is the score of the found motif, i.e. its distance to all sequences?\n",
    "\n",
    "Found motif is GTGACACATTATGTC with score 25.\n",
    "\n",
    "#### 4. How can we improve the performance of Gibbs sampling in order to be more reliable, faster, etc?\n",
    "\n",
    "We can impove reliability by repeating Gibbs sampling as we did in our implementation. The other option (\"phase shifts\") would be to shift candidate motifs when we get stuck in local optima with a hope of getting better results. We could also use information about amino acids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "225\n",
      "['GTTTTCCCA', 'GTCTTCGCG', 'TGTCTAGCT', 'AGTTACGCT', 'TGTTTCGGT', 'TTATTCACT', 'CTACGCGGA', 'AGTTTAACA', 'CTTCTCCCA', 'GTATTTCCT']\n"
     ]
    }
   ],
   "source": [
    "l = 9\n",
    "D = load_fasta('MotSamShort.fasta')\n",
    "DNAs = []\n",
    "for d, DNA in D.items():\n",
    "    DNAs.append(DNA)\n",
    "results = repeated_gibbs_sampling(DNAs, l, mutations=4)\n",
    "print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5. What results are produced by your program when searching for motif of length 9 with at most 4 mutations within sequences in the file  MotSamShort.fasta? How does the computation differ from the computation with the first set of sequences?  \n",
    "We have more motifs that fit the criteria of having at most 4 mutations and because length is only 9 we have more candidates. After repeating Gibbs sampling we don't obtain the same best motifs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
